### PR TITLE
Make associated_item_def_ids for traits use an unstable option to also return associated types for RPITITs

### PR DESCRIPTION
**Pull request by spastorino**
r? @compiler-errors

### CODE REVIEW

**Comment by spastorino** [comment]
I should have opened this one as draft. I'm going to add also the ability to encode rpitits based on the unstable option.

File: compiler/rustc_hir_analysis/src/astconv/mod.rs [hunk 1]
```diff
@@ -1440,6 +1441,14 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                             tcx.associated_items(pred.def_id())
                                 .in_definition_order()
                                 .filter(|item| item.kind == ty::AssocKind::Type)
+                                .filter(|item| {
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/astconv/mod.rs:1444]: What is this for?

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 2]
```diff
@@ -1130,8 +1130,18 @@ fn should_encode_trait_impl_trait_tys(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
 }
 
 // Return `false` to avoid encoding impl trait in trait, while we don't use the query.
```
Comment (compiler-errors) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:1132]: This comment needs to be modified to specify that it's specifically encoding the *new* RPITIT lowering (to assoc types).

File: compiler/rustc_hir_analysis/src/astconv/mod.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/astconv/mod.rs:1444]: Oh -- we should definitely be bailing out in the case that our trait has RPIT associated items and we're trying to make an object out of it. This is probably not the right approach?

Regardless, any modification like this needs some sort of comment justifying it.

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 3]
```diff
@@ -1130,8 +1130,18 @@ fn should_encode_trait_impl_trait_tys(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
...
-fn should_encode_fn_impl_trait_in_trait<'tcx>(_tcx: TyCtxt<'tcx>, _def_id: DefId) -> bool {
-    false
+fn should_encode_fn_impl_trait_in_trait<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> bool {
+    if tcx.sess.opts.unstable_opts.new_rpitit {
+        match tcx.def_kind(def_id) {
+            DefKind::Fn | DefKind::AssocFn => {
+                tcx.def_kind(tcx.parent(def_id)) == DefKind::Trait
+                    && tcx.hir().get_fn_output(def_id.expect_local()).is_some()
```
Comment (compiler-errors) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:1138]: Should we encode only if we know the trait has RPITITs? Is that easy to check?

File: compiler/rustc_trait_selection/src/traits/object_safety.rs [hunk 4]
```diff
@@ -157,6 +158,11 @@ fn object_safety_violations_for_trait(
                 .in_definition_order()
                 .filter(|item| item.kind == ty::AssocKind::Type)
                 .filter(|item| !tcx.generics_of(item.def_id).params.is_empty())
+                .filter(|item| {
+                    tcx.def_path(item.def_id).data.last().map_or(true, |def_path_data| {
+                        def_path_data.data != DefPathData::ImplTraitAssocTy
+                    })
+                })
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/object_safety.rs:165]: RPITITs are not object safe. Is this to prevent an ICE or what? It should probably turn into a new `ObjectSafetyViolation` kind.

File: compiler/rustc_ty_utils/src/assoc.rs [hunk 5]
```diff
@@ -21,9 +22,34 @@ pub fn provide(providers: &mut ty::query::Providers) {
 fn associated_item_def_ids(tcx: TyCtxt<'_>, def_id: DefId) -> &[DefId] {
     let item = tcx.hir().expect_item(def_id.expect_local());
     match item.kind {
-        hir::ItemKind::Trait(.., ref trait_item_refs) => tcx.arena.alloc_from_iter(
-            trait_item_refs.iter().map(|trait_item_ref| trait_item_ref.id.owner_id.to_def_id()),
-        ),
+        hir::ItemKind::Trait(.., ref trait_item_refs) => {
+            if tcx.sess.opts.unstable_opts.new_rpitit {
+                tcx.arena.alloc_from_iter(
```
Comment (compiler-errors) [inline compiler/rustc_ty_utils/src/assoc.rs:27]: This could use a block comment explaining what the strategy here is. Basically that we collect the RPITITs from the method's return type and turn them into items (but with more words).

File: compiler/rustc_ty_utils/src/assoc.rs [hunk 6]
```diff
@@ -193,7 +219,50 @@ fn associated_item_for_impl_trait_in_trait(
     let span = tcx.def_span(opaque_ty_def_id);
     let trait_assoc_ty =
         tcx.at(span).create_def(trait_def_id.expect_local(), DefPathData::ImplTraitAssocTy);
-    trait_assoc_ty.def_id()
+
```
Comment (compiler-errors) [inline compiler/rustc_ty_utils/src/assoc.rs:222]: Each one of these feedings should probably explain its correctness or why it's done the way it's done (or the FIXMEs need more words).

File: tests/rustdoc-ui/z-help.stdout [hunk 7]
```diff
@@ -98,6 +98,7 @@
     -Z      mir-pretty-relative-line-numbers=val -- use line numbers relative to the function in mir pretty printing
     -Z                       move-size-limit=val -- the size at which the `large_assignments` lint starts to be emitted
     -Z                       mutable-noalias=val -- emit noalias metadata for mutable references (default: yes)
+    -Z                            new-rpitit=val -- specify to use the new rpitit system that lowers return position impl trait in trait as associated types both in the trait and in the corresponding impl
```
Comment (compiler-errors) [inline tests/rustdoc-ui/z-help.stdout:101]: I'm not a fan of this `-Z` flag name. Maybe something like `-Z lower-impl-trait-in-trait-to-assoc-ty`? It's verbose but clearer.

And the description could be:

"modify the lowering strategy for `impl Trait` in traits so that they are lowered to generic associated types"

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 8]
```diff
@@ -1177,8 +1187,7 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
                 record!(self.tables.codegen_fn_attrs[def_id] <- self.tcx.codegen_fn_attrs(def_id));
             }
             if should_encode_visibility(def_kind) {
-                let vis =
-                    self.tcx.local_visibility(local_id).map_id(|def_id| def_id.local_def_index);
+                let vis = self.tcx.visibility(def_id).map_id(|def_id| def_id.index);
```
Comment (petrochenkov) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:1190]: Why is this change needed?
If the visibility turns out non-local here it means a bug, it won't be encoded correctly.
`local_visibility` should always be preferred if it's possible.

Comment (spastorino) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:1190]: Because for newly synthesized associated types coming from RPITITs we would need to feed the visibility value using the query system. Otherwise we would execute local_visibility using the synthesized def_id and it will crash trying to access the non existing HIR.

Comment (compiler-errors) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:1190]: `TyCtxt::local_visibility` calls the `TyCtxt::visibility` query, though, and you're feeding that. If you revert this change, how does it ICE?

File: compiler/rustc_hir_analysis/src/astconv/mod.rs (hunk 1 above)
Comment (spastorino) [reply compiler/rustc_hir_analysis/src/astconv/mod.rs]: Removed, it will be easier to justify with tests later.

File: compiler/rustc_metadata/src/rmeta/encoder.rs (hunk 2 above)
Comment (spastorino) [reply compiler/rustc_metadata/src/rmeta/encoder.rs]: Done

File: compiler/rustc_trait_selection/src/traits/object_safety.rs (hunk 4 above)
Comment (spastorino) [reply compiler/rustc_trait_selection/src/traits/object_safety.rs]: Removed, it will be easier to justify with tests later.

File: compiler/rustc_ty_utils/src/assoc.rs (hunk 5 above)
Comment (spastorino) [reply compiler/rustc_ty_utils/src/assoc.rs:30]: Done

File: compiler/rustc_ty_utils/src/assoc.rs (hunk 6 above)
Comment (spastorino) [reply compiler/rustc_ty_utils/src/assoc.rs:225]: Done

File: tests/rustdoc-ui/z-help.stdout (hunk 7 above)
Comment (spastorino) [reply tests/rustdoc-ui/z-help.stdout]: Done

File: compiler/rustc_metadata/src/rmeta/encoder.rs (hunk 8 above)
Comment (spastorino) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:1191]: Let's remove the change and we can introduce it later, some things are hard to justify now without any test.

File: compiler/rustc_metadata/src/rmeta/encoder.rs (hunk 3 above)
Comment (spastorino) [reply compiler/rustc_metadata/src/rmeta/encoder.rs:1139]: Well, I guess we can `associated_items_for_impl_trait_in_trait` query and check if not empty?.

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 9]
```diff
@@ -1129,9 +1129,20 @@ fn should_encode_trait_impl_trait_tys(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
...
-fn should_encode_fn_impl_trait_in_trait<'tcx>(_tcx: TyCtxt<'tcx>, _def_id: DefId) -> bool {
-    false
+// We only encode impl trait in trait when using `lower-impl-trait-in-trait-to-assoc-ty` unstable
+// option.
+fn should_encode_fn_impl_trait_in_trait<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> bool {
+    if tcx.sess.opts.unstable_opts.lower_impl_trait_in_trait_to_assoc_ty {
+        match tcx.def_kind(def_id) {
+            DefKind::Fn | DefKind::AssocFn => {
+                tcx.def_kind(tcx.parent(def_id)) == DefKind::Trait
+                    && tcx.hir().get_fn_output(def_id.expect_local()).is_some()
```
Comment (compiler-errors) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:1139]: Can you just call `opt_associated_item`, check the `container` and the `kind` there? I don't even think we need to check if the output of the fn is non-empty.

File: compiler/rustc_middle/src/query/mod.rs [hunk 10]
```diff
@@ -1146,6 +1153,15 @@ rustc_queries! {
         desc { |tcx| "looking up definition kind of `{}`", tcx.def_path_str(def_id) }
         cache_on_disk_if { def_id.is_local() }
         separate_provide_extern
+        feedable
+    }
+
+    /// The `opt_rpitit_info` query returns the pair of the def id of the function where the RPIT
+    /// is defined and the opaque def id if any.
+    query opt_rpitit_info(def_id: DefId) -> Option<(DefId, Option<DefId>)> {
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/query/mod.rs:1161]: Maybe this should be a struct or something

```
enum ImplTraitInTraitData {
    Trait { trait_fn_def_id: DefId },
    Impl { impl_fn_def_id: DefId, trait_opaque_item: DefId },
}
```

or whatever.

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 11]
```diff
@@ -1129,9 +1129,20 @@ fn should_encode_trait_impl_trait_tys(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
...
 
-// Return `false` to avoid encoding impl trait in trait, while we don't use the query.
-fn should_encode_fn_impl_trait_in_trait<'tcx>(_tcx: TyCtxt<'tcx>, _def_id: DefId) -> bool {
-    false
+// We only encode impl trait in trait when using `lower-impl-trait-in-trait-to-assoc-ty` unstable
+// option.
+fn should_encode_fn_impl_trait_in_trait<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> bool {
+    if tcx.sess.opts.unstable_opts.lower_impl_trait_in_trait_to_assoc_ty {
+        match tcx.def_kind(def_id) {
+            DefKind::Fn | DefKind::AssocFn => {
```
Comment (compiler-errors) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:1137]: `DefKind::Fn` is never going to have RPITITs.

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 12]
```diff
@@ -1129,9 +1129,17 @@ fn should_encode_trait_impl_trait_tys(tcx: TyCtxt<'_>, def_id: DefId) -> bool {
...
 }
 
-// Return `false` to avoid encoding impl trait in trait, while we don't use the query.
-fn should_encode_fn_impl_trait_in_trait<'tcx>(_tcx: TyCtxt<'tcx>, _def_id: DefId) -> bool {
-    false
+// We only encode impl trait in trait when using `lower-impl-trait-in-trait-to-assoc-ty` unstable
+// option.
+fn should_encode_fn_impl_trait_in_trait<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> bool {
+    if tcx.sess.opts.unstable_opts.lower_impl_trait_in_trait_to_assoc_ty {
+        match tcx.opt_associated_item(def_id) {
```
Comment (compiler-errors) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:1136]: hm, this probably should check for only functions too... I was thinking of something simpler than this.

```
if tcx.sess.opts.unstable_opts.lower_impl_trait_in_trait_to_assoc_ty
    && let Some(assoc_item) = tcx.opt_associated_item(def_id)
    && assoc_item.container == ty::AssocItemContainer::TraitContainer
    && assoc_item.kind == ty::AssocKind::Fn
{
    true
} else {
    false
}
```
